src/QMCDrivers/VMC/VMCBatched.cpp

```diff
@@ -2,68 +2,61 @@
+// File developed by: Peter Doak, doakpw@ornl.gov, Oak Ridge National Laboratory
 //
-// File created by: Jeongnim Kim, jeongnim.kim@gmail.com, University of Illinois at Urbana-Champaign
+// File refactored from: VMC.cpp
 //////////////////////////////////////////////////////////////////////////////////////
 
-
 #include "QMCDrivers/VMC/VMCBatched.h"
 #include "QMCDrivers/VMC/VMCUpdatePbyP.h"
 #include "QMCDrivers/VMC/VMCUpdateAll.h"
-#include "OhmmsApp/RandomNumberControl.h"
-#include "Message/OpenMP.h"
-#include "Message/CommOperators.h"
-#include "Utilities/RunTimeManager.h"
-#include <qmc_common.h>
-//#define ENABLE_VMC_OMP_MASTER
-#include "ADIOS/ADIOS_profile.h"
-#if !defined(REMOVE_TRACEMANAGER)
-#include "Estimators/TraceManager.h"
-#else
-typedef int TraceManager;
-#endif
 
 namespace qmcplusplus
 {
-/// Constructor.
-VMCBatched::VMCBatched(MCPopulation& pop,
+/** Constructor maintains proper ownership of input parameters
+   */
+VMCBatched::VMCBatched(QMCDriverInput& qmcdriver_input,
+                       VMCDriverInput& input,
+                       MCPopulation& pop,
                        TrialWaveFunction& psi,
                        QMCHamiltonian& h,
                        WaveFunctionPool& ppool,
                        Communicate* comm)
-    : QMCDriverNew(pop, psi, h, ppool, comm), UseDrift("yes")
+    : QMCDriverNew(qmcdriver_input, pop, psi, h, ppool, comm), vmcdriver_input_(input)
 {
-  RootName = "vmc";
-  QMCType  = "VMCBatched";
-  qmc_driver_mode.set(QMC_UPDATE_MODE, 1);
-  qmc_driver_mode.set(QMC_WARMUP, 0);
-  m_param.add(UseDrift, "useDrift", "string");
-  m_param.add(UseDrift, "usedrift", "string");
-  m_param.add(UseDrift, "use_drift", "string");
+  // qmc_driver_mode.set(QMC_UPDATE_MODE, 1);
+  // qmc_driver_mode.set(QMC_WARMUP, 0);
+}
 
-  prevSteps               = nSteps;
-  prevStepsBetweenSamples = nStepsBetweenSamples;
+VMCBatched::IndexType VMCBatched::calc_default_local_walkers()
+{
+  IndexType rw = vmcdriver_input_.get_requested_walkers_per_rank();
+  if (rw < num_crowds_)
+    rw = num_crowds_;
+  walkers_per_crowd_      = (rw % num_crowds_) ? rw / num_crowds_ + 1 : rw / num_crowds_;
+  IndexType local_walkers = walkers_per_crowd_ * num_crowds_;
+  population_.set_num_local_walkers(local_walkers);
+  population_.set_num_global_walkers(local_walkers * population_.get_num_ranks());
+  if (rw != vmcdriver_input_.get_requested_walkers_per_rank())
+    app_warning() << "VMCBatched driver has adjusted walkers per rank to: " << local_walkers << '\n';
+
+  if (vmcdriver_input_.get_samples() >= 0 || vmcdriver_input_.get_samples_per_thread() >= 0 ||
+      vmcdriver_input_.get_steps_between_samples() >= 0)
+    app_warning() << "VMCBatched currently ignores samples and samplesperthread\n";
+
+  // TODO: Simplify samples, samples per thread etc in the unified driver
+  // see login in original VMC.cpp
+  return local_walkers;
```

Comment:
l thnk the current implementation calc_default_local_walkers is OK for experiment purpose but needs to be generalized.

Reply:
As far as I can tell only VMC behaved like this.  Which is why I stuck this out in the derived class. If we can remove the "weird" behavior and get rid of it that would be good.